### PR TITLE
Websocket worker example

### PR DESCRIPTION
**Pull request by midokura-silvia**
This pull request contains an example implementation of a federated training using websockets and federated averaging. The training on the workers is currently run sequentially (no scheduler implemented).

### CODE REVIEW

File: examples/tutorials/Advanced/Websockets - Federated training/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+The scripts in this folder let you execute a federated training via three websocket connections.
+
+The script start_websocket_servers.py will start the Websocket server workers for Alice, Bob and Charlie.
+```
+$ python start_websocket_servers.py
+```
+
+The training is then started by running the script run_websocket_client.py:
+```
+$ python run_websocket_client.py
```
Comment (robert-wagner) [inline examples/tutorials/Advanced/Websockets - Federated training/README.md]: Any way that we could create a version of this in a notebook so that people trying to use it can iterate step by step?

File: examples/tutorials/Advanced/Websockets - Federated training/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+```
+This script
+ * loads the MNIST dataset,
+ * distributes it onto the three workers
```
Comment (robert-wagner) [inline examples/tutorials/Advanced/Websockets - Federated training/README.md]: Could we do this step by intitializing each of alice, bob and charlie with the data parameter. That seems to be much more in line with how fl will work in the wild

File: examples/tutorials/Advanced/Websockets - Federated training/run_websocket_server.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    p.start()
+    return p
+
+
+parser = argparse.ArgumentParser(description="Run websocket server worker.")
+parser.add_argument(
+    "--port", "-p", type=int, default=8777, help="port number of the websocket server worker"
+)
+parser.add_argument("--host", type=str, default="localhost", help="host for the connection")
+parser.add_argument("--id", default="alice", help="name (id) of the websocket server worker")
```
Comment (robert-wagner) [inline examples/tutorials/Advanced/Websockets - Federated training/run_websocket_server.py]: This probably shouldn't have a default as that can lead to conflicts

File: examples/tutorials/Advanced/Websockets - Federated training/run_websocket_server.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        server.start()
+
+    p = Process(target=target)
+    p.start()
+    return p
+
+
+parser = argparse.ArgumentParser(description="Run websocket server worker.")
+parser.add_argument(
+    "--port", "-p", type=int, default=8777, help="port number of the websocket server worker"
```
Comment (robert-wagner) [inline examples/tutorials/Advanced/Websockets - Federated training/run_websocket_server.py]: This should probably not have a default as that can lead to conflicts

File: syft/frameworks/torch/federated/dataloader.py [hunk 5]
```diff
@@ -110,6 +110,50 @@ def stop(self):
...
+        # Assign the worker to invoke
+        self.worker = loader.workers[worker_idx]
+
+        # The function used to stack all samples together
+        self.collate_fn = loader.collate_fn
+
+        # Create a sample iterator for each worker
+        self.sample_iter = iter(loader.batch_samplers[self.worker])
+
+    # def __len__(self):
```
Comment (robert-wagner) [inline syft/frameworks/torch/federated/dataloader.py]: why is this commented out

File: syft/frameworks/torch/federated/dataloader.py [hunk 6]
```diff
@@ -110,6 +110,50 @@ def stop(self):
         raise StopIteration
 
 
+class _DataLoaderOneWorkerIter(object):
+    r"""Iterates once over the worker's dataset, as specified by its sampler"""
```
Comment (robert-wagner) [inline syft/frameworks/torch/federated/dataloader.py]: does this need to be a raw string literal?

File: syft/workers/websocket_client.py [hunk 7]
```diff
@@ -32,20 +36,36 @@ def __init__(
...
+        self.ws.send(str(binascii.hexlify(message)))
+        response = binascii.unhexlify(self.ws.recv()[2:-1])
+        return response
+
     def _recv_msg(self, message: bin) -> bin:
         """Forwards a message to the WebsocketServerWorker"""
 
-        self.ws.send(str(binascii.hexlify(message)))
-        response = binascii.unhexlify(self.ws.recv()[2:-1])
+        response = self._receive_action(message)
```
Comment (robert-wagner) [inline syft/workers/websocket_client.py]: This will fail or hang if the websocket is not connected. What is the check after this actually doing?

File: test/torch/federated/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+import torch
+from syft import TorchHook
+
+from syft.frameworks.torch.federated import utils
+from syft.frameworks.torch import federated
+
+hook = TorchHook(torch)
+
+
+class DataLoaderTestCase(unittest.TestCase):
```
Comment (robert-wagner) [inline test/torch/federated/test_utils.py]: Format this in pytest style

File: test/torch/federated/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    def setUp(self):
```
Comment (robert-wagner) [inline test/torch/federated/test_utils.py]: This is handled by the conftest workers

File: test/torch/federated/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+import torch as th
+import syft as sy
+import torch
+from syft import TorchHook
+
+from syft.frameworks.torch.federated import utils
+from syft.frameworks.torch import federated
+
+hook = TorchHook(torch)
```
Comment (robert-wagner) [inline test/torch/federated/test_utils.py]: this is handled by the conftest hook

File: examples/tutorials/Advanced/Websockets - Federated training/README.md (hunk 1 above)
Comment (midokura-silvia) [reply examples/tutorials/Advanced/Websockets - Federated training/README.md]: I thought of it, but debugging a notebook is more difficult than putting breakpoints in an editor as pycharm. 
The script starting the websocket servers will not go away, as starting the processes from a notebook does not work (see discussion in the slack channel).
I will see what makes sense in a notebook, as it shouldn't be too long neither.

File: syft/frameworks/torch/federated/dataloader.py (hunk 5 above)
Comment (midokura-silvia) [reply syft/frameworks/torch/federated/dataloader.py]: That is a left-over of some tests. I will un-comment it.

File: syft/frameworks/torch/federated/dataloader.py (hunk 6 above)
Comment (midokura-silvia) [reply syft/frameworks/torch/federated/dataloader.py]: No, I don't think so. I will remove the r-prefix from all the docstrings in the file.

File: syft/workers/websocket_client.py (hunk 7 above)
Comment (midokura-silvia) [reply syft/workers/websocket_client.py]: _receive_action will not fail, if the websocket is not connected. but the response will be empty. 
The check on self.ws.connected has to be after a self.ws.recv() call as otherwise the variable is not updated. 
So the steps are: 1) try to send/receive, 2) check if connection is ok, if not 3) reconnect and send/receive again. 4) check if connection is ok now, if not fail.

File: test/torch/federated/test_utils.py (hunk 8 above)
Comment (midokura-silvia) [reply test/torch/federated/test_utils.py]: will do

**Comment by iamtrask** [comment]
Hey @midokura-silvia!! This is some great work! Are you in PySyft's slack? I'd love to have a chat.

File: examples/tutorials/Advanced/Websockets - Federated training/README.md (hunk 1 above)
Comment (midokura-silvia) [reply examples/tutorials/Advanced/Websockets - Federated training/README.md]: I added a notebook version of the example.

File: examples/tutorials/Advanced/Websockets - Federated training/README.md (hunk 2 above)
Comment (midokura-silvia) [reply examples/tutorials/Advanced/Websockets - Federated training/README.md]: This could be a separate example, where the focus is on setting up the remote data in a realistic way. My suggestion is to do a separate pull request for this.

File: examples/tutorials/Advanced/Websockets - Federated training/README.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# Federated learning using websockets - MNIST example
```
Comment (robert-wagner) [inline examples/tutorials/Advanced/Websockets - Federated training/README.md]: Could you make the name of this folder more command line friendly (ie no spaces

Comment (midokura-silvia) [reply examples/tutorials/Advanced/Websockets - Federated training/README.md]: Would the name "websockets-example-MNIST" be good?
Shall we change the "Advanced" folder to "advanced" to align it as well?

**Comment by iamtrask** [comment]
Excellent work Silvia! And Also excellent reviewing Bobby!
